### PR TITLE
Update implementation for ContextValue

### PR DESCRIPTION
**Pull request by gabizou**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1648) | **SpongeCommon**

This updates the implementation, borrowing the implementation of previously `NamedCauseFilter`.

### CODE REVIEW

File: src/test/java/org/spongepowered/common/event/listener/ContextValueCauseListener.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package org.spongepowered.common.event.listener;
```
Comment (kashike) [inline src/test/java/org/spongepowered/common/event/listener/ContextValueCauseListener.java:1]: license header

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/event/filter/delegate/NamedCauseFilterSourceDelegate.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+public class NamedCauseFilterSourceDelegate extends CauseFilterSourceDelegate {
+
+    private final ContextValue anno;
+
+    public NamedCauseFilterSourceDelegate(ContextValue anno) {
+        this.anno = anno;
```
Comment (Deamon5550) [inline src/main/java/org/spongepowered/common/event/filter/delegate/NamedCauseFilterSourceDelegate.java:56]: Assert here that the id in the annotation is a valid key and throw an exception if not.

File: src/main/java/org/spongepowered/common/event/filter/delegate/NamedCauseFilterSourceDelegate.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @Override
+    protected void insertCauseCall(MethodVisitor mv, Parameter param, Class<?> targetType) {
+        mv.visitMethodInsn(INVOKEVIRTUAL, Type.getInternalName(Cause.class), "getContext", "()" + Type.getDescriptor(EventContext.class), false);
```
Comment (Deamon5550) [inline src/main/java/org/spongepowered/common/event/filter/delegate/NamedCauseFilterSourceDelegate.java:61]: Throw an exception here if the parameter type does not match the type of the key.

File: src/main/java/org/spongepowered/common/event/filter/delegate/NamedCauseFilterSourceDelegate.java (hunk 2 above)
Comment (Deamon5550) [reply src/main/java/org/spongepowered/common/event/filter/delegate/NamedCauseFilterSourceDelegate.java:56]: This filter delegate should be creating a field in the filter holding the `EventContextKey` which is fetched from the registry by id when the filter is initialized (look at the subtype filter for how creating a field and putting stuff into the initializer works)

File: src/test/java/org/spongepowered/common/event/listener/ContextValueCauseListener.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    @Listener
+    public void namedCauseListenerInclude(EventFilterTest.SubEvent event, @ContextValue(value = EventContextKeys.Named.OWNER_NAME, typeFilter=User.class) Object state) {
+        this.namedCauseListenerCalledInc = true;
+    }
+
+    @Listener
+    public void namedCauseListenerExclude(EventFilterTest.SubEvent event, @ContextValue(value = EventContextKeys.Named.OWNER_NAME, typeFilter=User.class, inverse = true) Object state) {
+        this.namedCauseListenerCalledEx = true;
```
Comment (Deamon5550) [inline src/test/java/org/spongepowered/common/event/listener/ContextValueCauseListener.java:28]: this needs to be inverted since you are testing for false

**Review by Deamon5550 — CHANGES_REQUESTED** [review]

**Comment by ryantheleach** [comment]
Closed by @Zidane , See API PR.

**Comment by Deamon5550** [comment]
@ryantheleach this PR is still needed as the `ContextValue` filter is still not implemented. The api pr had adjacent changes which is what @Zidane was rejecting.

**Pull request closed without merging**